### ISSUE TITLE
cmake fails to build on macOS Sierra with GCC@6.2.0

### ISSUE THREAD

**Issue by davydden** [LLNL/spack#1847]
subj, during `bootstrap(*options)`. In Spack logs:

```
In file included from /usr/include/Availability.h:176:0,
                 from /usr/include/inttypes.h:224,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Utilities/KWIML/include/kwiml/int.h:194,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Utilities/cm_kwiml.h:22,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmStandardIncludes.h:34,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmObject.h:15,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmCommand.h:15,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmAddCustomCommandCommand.h:15,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmAddCustomCommandCommand.cxx:12,
                 from /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/cmBootstrapCommands1.cxx:16:
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFDateFormatter.h:53:34: error: ‘introduced’ was not declared in this scope
     kCFISO8601DateFormatWithYear API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0)) = (1UL << 0),
```

and in boostraping logs:

```
/private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/kwsys/kwsysPlatformTestsCXX.cxx: In function ‘int main()’:
/private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/kwsys/kwsysPlatformTestsCXX.cxx:189:13: error: ‘environ’ was not declared in this scope
   char* e = environ[0];
             ^~~~~~~
Test failed to compile
Try: /Users/davydden/spack/lib/spack/env/gcc/g++
Line: /Users/davydden/spack/lib/spack/env/gcc/g++  -DTEST_KWSYS_STL_HAS_WSTRING /private/var/folders/5k/sqpp24tx3ylds4fgm13pfht00000gn/T/davydden/spack-stage/spack-stage-nz7Rvi/cmake-3.6.1/Source/kwsys/kwsysPlatformTestsCXX.cxx -o cmake_bootstrap_15628_test
```

**Comment by kiwifb** [comment]
Classic Apple problem. You see clang does objective C as well as C by default. So the Apple upstream doesn't see a problem putting objective C code in their system headers without any language guard. 
Once some of those polluted headers get pulled in gcc will fail. It looks to me like the problems are more acute in Sierra than they were in El Capitan, Availability.h gets pulled by a number of standard headers. It is getting very difficult to use gcc on OS X and Apple doesn't care.

**Comment by davydden** [comment]
thanks for the explanation. For what it's worth, i will try reporting it to Apple as a bug.

**Comment by davydden** [comment]
reported upstream...

**Comment by kiwifb** [comment]
Been trolling old sagemath bugs. Is `MACOSX_DEPLOYMENT_TARGET` set anywhere in spack? Could you try to unset it?

**Comment by kiwifb** [comment]
Culprit is not the fact that clang is an objective C compiler, although it may contribute. From https://trac.sagemath.org/ticket/18254#comment:60:

As for upstream, I've gotten some good response on the osx-dev forums:

​https://devforums.apple.com/message/1128923#1128923

tl;dr: Fixing this upstream involves enhancements to one of both gcc and clang:

clang: the macros in Availability.h need an "ENUM" variant (like we have for the CF and NS availability macros) to cover this case
gcc: implement clang's extension ​http://clang.llvm.org/docs/LanguageExtensions.html#attributes-on-enumerators

**Comment by healther** [comment]
@davydden @kiwifb Is there a solution or a workaround to this problem? (I can't follow your forum link, it redirects to the welcome page, even after signing in)
Or alternatively, can I make spack just use a system cmake version (that I would install manually?)

**Comment by adamjstewart** [comment]
> Or alternatively, can I make spack just use a system cmake version (that I would install manually?)

http://spack.readthedocs.io/en/latest/build_settings.html#external-packages

**Comment by kiwifb** [comment]
I don't have access to the apple forum either. For `cmake` it doesn't matter the compiler you build it with and it should build just fine with the system's `clang`. That's what `cmake` upstream does.

**Comment by adamjstewart** [comment]
@davydden @mathstuf Did we ever get this fixed?

**Comment by davydden** [comment]
I don't think so

**Comment by davydden** [comment]
I am closing this as I don't think we can do much but to report the issue to Apple and hope it is/will be fixed.
